File: 10.4_MyPhotos/.idea/app/src/main/java/com/example/a55myphotos/MainActivity.java
package com.example.a55myphotos;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.example.a55myphotos.View.GetPosition;
import com.example.a55myphotos.View.PhotoAdapter;
import com.example.a55myphotos.View.SinglePhotoActivity;
import com.example.a55myphotos.model.Photo;
import com.example.a55myphotos.network.GetDataService;
import com.example.a55myphotos.network.RetrofitClient;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity implements Serializable {
    private RecyclerView recyclerView;
    public List<Photo> photos = new ArrayList<>();
    public Call<List<Photo>> callPhotos;
    protected Response<List<Photo>> response;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView = findViewById(R.id.recyclerview);
        GetDataService getDataService = RetrofitClient.getRetrofit().create(GetDataService.class);
        callPhotos = getDataService.getPhotos();
        retrofitCall();
    }

    public void createRecycler() {
        PhotoAdapter adapter = new PhotoAdapter(photos);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(adapter);
        adapter.setGetPosition(new GetPosition() {
            @Override
            public void onClick(int position) {

                Intent intent = new Intent(getApplicationContext(), SinglePhotoActivity.class);
                intent.putExtra("MyObject", photos.get(position));
                startActivity(intent);
            }
        });
    }

    public void retrofitCall() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    response = callPhotos.execute();
                } catch (IOException e) {
                    e.printStackTrace();
                }
                photos = response.body();
                recyclerView.post(new Runnable() {
                    @Override
                    public void run() {
                        createRecycler();
                    }
                });
            }
        }).start();
    }
}

File: 5.5 MyPhotosRetrofit-KotlinVersion/app/src/main/java/com/example/a55myphotosretrofit_kotlinversion/MainActivity.kt
package com.example.a55myphotosretrofit_kotlinversion

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.a55myphotosretrofit_kotlinversion.model.Photo
import com.example.a55myphotosretrofit_kotlinversion.network.GetDataService
import com.example.a55myphotosretrofit_kotlinversion.network.RetrofitClient
import com.example.a55myphotosretrofit_kotlinversion.view.PhotoAdapter
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        val getDataService = RetrofitClient.retrofit?.create(GetDataService::class.java)
        val callPhotos: Call<List<Photo>>? = getDataService?.getPhotos()

        callPhotos?.enqueue(object : Callback<List<Photo>> {

            override fun onResponse(call: Call<List<Photo>>, response: Response<List<Photo>>) {
                response.body()?.let { callPhotos ->
                    for (photo in callPhotos) {

                    }
                    createRecycler(callPhotos)
                }
            }

            override fun onFailure(call: Call<List<Photo>>, t: Throwable) {

            }
        })
    }


    fun createRecycler(photos: List<Photo>) {
        val adapter = PhotoAdapter((photos)) { adapterPosition ->
            Toast.makeText(this, "Item in $adapterPosition is clicked", Toast.LENGTH_SHORT).show()

        }
        recyclerview.layoutManager = LinearLayoutManager(this)
        recyclerview.adapter = adapter
    }
}

